### PR TITLE
chore: Upstream merge as of 01122021

### PR DESCRIPTION
**Pull request by mukhoakash**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: pkg/azureutils/azure_disk_utils.go [hunk 1]
```diff
@@ -206,13 +206,15 @@ func CreateValidDiskName(volumeName string, usedForLabel bool) string {
 }
 
 // GetCloudProviderFromClient get Azure Cloud Provider
-func GetCloudProviderFromClient(kubeClient *clientset.Clientset, secretName string, secretNamespace string, userAgent string) (*azure.Cloud, error) {
```
Comment (landreasyan) [inline pkg/azureutils/azure_disk_utils.go:209]: We shouldn't override our changes in `GetCloudProviderFromClient`. The changes from master should be ignored for this function.

**Review by landreasyan — CHANGES_REQUESTED** [review]
Since the old/upstream implementation of  GetCloudProvider affects our performance, we need to take our changes for that function during the merge.

File: go.mod [hunk 2]
```diff
@@ -24,18 +24,18 @@ require (
...
+	k8s.io/client-go v0.22.4
+	k8s.io/cloud-provider v0.22.4
+	k8s.io/code-generator v0.22.4
+	k8s.io/component-base v0.22.4
+	k8s.io/component-helpers v0.22.4
 	k8s.io/klog v1.0.0
 	k8s.io/klog/v2 v2.10.0
-	k8s.io/kube-scheduler v0.22.3
+	k8s.io/kube-scheduler v0.22.4
 	k8s.io/kubernetes v1.22.3
```
Comment (edreed) [inline go.mod:37]: I wonder if we should bump `k8s.io/kubernetes` up to v1.22.4 as well.

File: pkg/azureutils/azure_disk_utils.go (hunk 1 above)
Comment (edreed) [reply pkg/azureutils/azure_disk_utils.go:209]: Well, except that we'll need to support the `allowEmptyCloudConfig` parameter, though I'm not sure why this was added in the first place. I'm fine with ignoring it for now as long as we file an issue to follow up on what the intended use case is for it.

File: pkg/azuredisk/controllerserver.go [hunk 3]
```diff
@@ -890,9 +891,11 @@ func (d *Driver) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshotRequ
 			}
 		case consts.ResourceGroupField:
 			resourceGroup = v
+		case consts.LocationField:
```
Comment (edreed) [inline pkg/azuredisk/controllerserver.go:894]: Need to add support for the "location" parameter.

**Review by edreed — CHANGES_REQUESTED** [review]
In addition to comments, please resolve test failures before merging.

File: pkg/azureutils/azure_disk_utils.go (hunk 1 above)
Comment (mukhoakash) [reply pkg/azureutils/azure_disk_utils.go:209]: As @landreasyan suggested to keep the existing implementation of `GetCloudProviderFromClient`, I will remove the dependency on `allowEmptyCloudConfig` for now until we reach a resolution.

File: pkg/azuredisk/controllerserver.go (hunk 3 above)
Comment (mukhoakash) [reply pkg/azuredisk/controllerserver.go:894]: Replicated to `CloudProvisioner`

File: go.mod (hunk 2 above)
Comment (mukhoakash) [reply go.mod]: Done.

File: charts/index.yaml [hunk 4]
```diff
@@ -12,7 +12,25 @@ entries:
...
-    created: "2021-11-15T22:22:06.6033492-08:00"
+    created: "2021-12-01T12:44:49.24528479-08:00"
+    description: Azure disk Container Storage Interface (CSI) Storage Plugin
+    digest: 147b31df2989e7b116e6d289ecdb817670d739dc1c399f42594c6fa0a4a1547f
+    name: azuredisk-csi-driver
+    urls:
+    - https://raw.githubusercontent.com/abhisheksinghbaghel/azuredisk-csi-driver/master/charts/latest/azuredisk-csi-driver-v1.10.0.tgz
+    version: v1.10.0
+  - apiVersion: v1
+    appVersion: latest
```
[the remark this answers is not in the document]
Comment (mukhoakash) [reply charts/index.yaml]: Pulled again from upstream, this got resolved.

**Review by edreed — APPROVED** [review]
The e2e tests are failing due to missing btrfs components. They'll need to be added to our Dockerfile: [btrfs-progs](https://pkgs.alpinelinux.org/packages?name=btrfs-progs&branch=v3.14).

```log
I1206 09:23:46.075946       1 mount_linux.go:425] Disk "/dev/disk/azure/scsi1/lun0" appears to be unformatted, attempting to format as type: "btrfs" with options: [/dev/disk/azure/scsi1/lun0]
E1206 09:23:46.075988       1 mount_linux.go:431] format of disk "/dev/disk/azure/scsi1/lun0" failed: type:("btrfs") target:("/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-8aaf3b14-e1c8-47df-a70e-d59729b0bac4/globalmount") options:("defaults") errcode:(executable file not found in $PATH) output:() 
E1206 09:23:46.076012       1 utils.go:101] GRPC error: rpc error: code = Internal desc = could not format "/dev/disk/azure/scsi1/lun0"(lun: '\x00'), and mount it at "/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-8aaf3b14-e1c8-47df-a70e-d59729b0bac4/globalmount"
```

**Review by landreasyan — APPROVED** [review]

**Pull request merged**